### PR TITLE
Add repair unit tests/v1

### PR DESCRIPTION
**Pull request by mikolajsieluzycki**
This patch series splits up parts of repair pipeline to allow unit testing 
various bits of code without having to run full dtest suite. The reason why
repair pipeline has no unit tests is that by definition repair requires multiple 
nodes, while unit test environment works only for a single node.

However, it is possible to explicitly define interfaces between various parts of the 
pipeline, inject dependencies and test them individually. This patch series is focused
on taking repair_rows_on_wire (frozen mutation representation of changes coming from 
another node) and flushing them to an sstable.

The commits are split into the following parts:
- pulling out classes to separate headers so that they can be included (potentially indirectly) from the test,
- pulling out repair_meta::to_repair_rows_list and part of repair_meta::flush_rows_in_working_row_buf so that they can be tested,
- refactoring repair_writer so that the actual writing logic can be injected as dependency,
- creating the unit test.

tests: unit(dev), dtest(incremental_repair_test, read_repair_test, repair_additional_test, repair_test)

### CODE REVIEW

File: repair/repair_hash.hh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+using repair_hash_set = absl::btree_set<repair_hash>;
+
+class repair_hasher {
+    uint64_t _seed;
+    schema_ptr _schema;
+public:
+    repair_hasher(uint64_t seed, schema_ptr s)
+        : _seed(seed)
+        , _schema(s)
```
Comment (raphaelsc) [inline repair/repair_hash.hh:45]: _schema(std::move(s))

File: repair/row_level.cc [hunk 2]
```diff
@@ -552,12 +552,12 @@ void flush_rows(schema_ptr s, std::list<repair_row>& rows, lw_shared_ptr<repair_
     auto cmp = position_in_partition::tri_compare(*s);
     lw_shared_ptr<mutation_fragment> last_mf;
     lw_shared_ptr<const decorated_key_with_hash> last_dk;
+    writer->start_streaming(db, sys_dist_ks, view_update_generator);
     for (auto& r : rows) {
         thread::maybe_yield();
         if (!r.dirty_on_master()) {
             continue;
         }
-        writer->create_writer(db, sys_dist_ks, view_update_generator);
```
Comment (raphaelsc) [inline repair/row_level.cc]: perhaps you should move it in a separate patch explaining the motivation.

File: repair/repair_writer.hh [hunk 3]
```diff
@@ -57,25 +94,29 @@ public:
         return _sem;
     }
 
+    schema_ptr schema() {
```
Comment (raphaelsc) [inline repair/repair_writer.hh:97]: const qualified

**Comment by raphaelsc** [comment]
The series made it very easy even for a non repair expert to review it. Congrats.

File: repair/repair_sync_boundary.hh [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+#pragma once
+#include "dht/i_partitioner.hh"
+#include "position_in_partition.hh"
+
+// Represent a position of a mutation_fragment read from a flat mutation
+// reader. Repair nodes negotiate a small range identified by two
+// repair_sync_boundary to work on in each round.
+struct repair_sync_boundary {
```
Comment (nyh) [inline repair/repair_sync_boundary.hh:16]: I understood why you wanted to move things out of the source file (repair.cc) and into header files to make them testable, but I didn't understand why you wanted to move things from repair.hh into separate header files. What does this movement get us?
I'm not strongly opposed to this move - it's nothing wrong with it per se - it's just that I don't understand the gain.

File: repair/repair_writer.hh [hunk 5]
```diff
@@ -45,6 +45,19 @@ public:
             , _reason(reason) {
     }
 
+    void start_streaming(sharded<replica::database>& _db, sharded<db::system_distributed_keyspace>& _sys_dist_ks, sharded<db::view::view_update_generator>& _view_update_generator) {
+        create_writer(_db, _sys_dist_ks, _view_update_generator);
```
Comment (nyh) [inline repair/repair_writer.hh]: What's the point of making the "create_writer" function private, and then writing a new function "start_streaming" that is public and just calls it? Why not keep the original "create_writer" public?

File: test/boost/repair.cc [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        std::vector<mutation> muts = gen(100);
+        auto m = make_lw_shared<replica::memtable>(s);
+        repair_rows_on_wire input;
+
+        for (mutation& mut : muts) {
+            partition_key pk = mut.key();
+            auto m2 = make_lw_shared<replica::memtable>(s);
+            m->apply(mut);
+            m2->apply(mut);
+            flat_mutation_reader reader = downgrade_to_v1(m2->make_flat_reader(s, permit));
```
Comment (nyh) [inline test/boost/repair.cc:114]: @denesb are we planning to remove this "downgrade_to_v1" stuff in the near future? If we are, let's not write new tests that uses it.

File: test/boost/repair.cc [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    virtual void start_streaming(lw_shared_ptr<repair_writer> writer) override {}
+};
+
+lw_shared_ptr<repair_writer> make_test_repair_writer(schema_ptr schema, reader_permit permit, mutation_fragment_queue mq) {
+    return make_lw_shared<repair_writer>(std::move(schema), std::move(permit), std::make_unique<test_repair_writer_impl>(std::move(mq)));
+}
+
+SEASTAR_TEST_CASE(flush_repair_rows_on_wire_to_sstable) {
```
Comment (nyh) [inline test/boost/repair.cc:96]: If this test is a unit-test for a specific class or function, please add a comment with a concrete explanation of what function or functions this test case checks. There's a lot (over 130 lines) of code here and it's hard to understand what is being tested and what is just an implementation detail of the test.

**Review by nyh — COMMENTED** [review]
Looks mostly good, I left a few minor comments.
@asias please review this series as it mostly touches code that you are most familiar with and wrote most if not all of it yourself.

File: test/boost/repair.cc (hunk 7 above)
Comment (avikivity) [inline test/boost/repair.cc:96]: Why flush to disk? Can't we compare two readers without going through an sstable? Asking in the interest of test speed.

**Comment by avikivity** [comment]
How long does the test run in dev/debug modes?

File: test/boost/repair.cc (hunk 7 above)
Comment (nyh) [reply test/boost/repair.cc:96]: This is why it's important to add a comment on the test stating exactly which function or class or whatever it acts as a unit test for. If we're testing a function whose purpose is to write mutations to sstables, you can't test it without allowing this function to write the sstables. But if the tests is supposed to test repair's function which compares two mutation streams (and if I understand correctly, this is NOT what this test is about) then you could write the test for it without involving sstables.

Something I didn't understand, by the way, is why does the repair code have special code which writes mutations to sstables. Don't we have other code which writes mutations to sstables - and why would we want to test such code as part of repair's unit tests?

File: repair/repair_writer.hh (hunk 5 above)
Comment (asias) [reply repair/repair_writer.hh]: I agree. This new start_streaming is also confusing.  It is not starting streaming but just create a write.

File: repair/row_level.cc (hunk 2 above)
Comment (asias) [reply repair/row_level.cc]: Yes. This changes the logic of the code. Before if the rows are empty, the writer is not created. Now, it creates the write unconditionally.

Comment (asias) [reply repair/row_level.cc]: Let's not do it for now in this series. The optimization is minor.

File: repair/row_level.cc [hunk 8]
```diff
@@ -1169,6 +1169,10 @@ class repair_meta {
         } else {
             return;
         }
+        flush_rows(_schema, _working_row_buf, _repair_writer, _db, _sys_dist_ks, _view_update_generator);
+    }
+
+    static void flush_rows(schema_ptr _schema, std::list<repair_row>& _working_row_buf, lw_shared_ptr<repair_writer>& _repair_writer, seastar::sharded<replica::database>& _db, seastar::sharded<db::system_distributed_keyspace>& _sys_dist_ks, seastar::sharded<db::view::view_update_generator>& _view_update_generator) {
```
Comment (asias) [inline repair/row_level.cc]: Let's say this patch introduces a new helper function flush_rows in the commit log and you want to make the helper a free function later. It took me a while to figure out how you split ;-)

File: test/boost/repair.cc [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        , _permit(std::move(permit))
+        , _sst(std::move(sst))
+        , _cfg(std::move(cfg))
+    {
+        _consumer = std::make_unique<upgrading_consumer<std::function<void(mutation_fragment_v2&&)>>>
+            (*_schema, _permit, [this](mutation_fragment_v2&& mf) mutable {
+                _fragments.push_back(std::move(mf));
+            });
+    }
+
```
Comment (asias) [inline test/boost/repair.cc:52]: minor: extra space here.

File: test/boost/repair.cc [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    virtual future<> wait_for_writer_done() override {
+        // Pushing EOS to the underlying mq waits until the sstable is written
+        return make_ready_future();
+    }
+
+    virtual mutation_fragment_queue& queue() override {
+        return _queue;
+    }
+
+    virtual void start_streaming(lw_shared_ptr<repair_writer> writer) override {}
```
Comment (asias) [inline test/boost/repair.cc:89]: Could you add comments about why we do nothing in this implementation?

File: test/boost/repair.cc [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+        for (mutation& mut : muts) {
+            partition_key pk = mut.key();
+            auto m2 = make_lw_shared<replica::memtable>(s);
+            m->apply(mut);
+            m2->apply(mut);
+            flat_mutation_reader reader = downgrade_to_v1(m2->make_flat_reader(s, permit));
+            std::list<frozen_mutation_fragment> mfs;
+            reader.consume_pausable([&input, s, &mfs](mutation_fragment mf) {
+                if ((mf.is_partition_start() && !mf.as_partition_start().partition_tombstone()) || mf.is_end_of_partition()) {
```
Comment (asias) [inline test/boost/repair.cc:117]: Let's add comments why we add the shortcut here.

**Comment by asias** [comment]
Looks mostly good. I added some comments. Thanks @mikolajsieluzycki !

File: test/boost/repair.cc (hunk 6 above)
Comment (denesb) [reply test/boost/repair.cc:114]: This series is part of an effort to move repair off v1 readers. We might be forced to work with v1 readers here for the time being. Of course if possible, we should write all code with v2 readers.

File: test/boost/repair.cc (hunk 7 above)
Comment (denesb) [reply test/boost/repair.cc:96]: What kind of special code do you mean? Repair uses the standard tools we have for writing a fragment stream into an sstable. The only special code it has is the multishard writer, which makes sure each partitions gets written to the shard which owns it. Repair runs on a single shard on each node, incoming data will usually cover any or all shards.

Comment (nyh) [reply test/boost/repair.cc:96]: I guess I just lost track of what exactly this unit-test actually tests. Not the full list of what the test **does** (which describes a lot of code included in the test itself, not in Scylla), but the actual Scylla function that is being **tested**. I thought that some sort of sstable-writer was being tested, otherwise I share Avi's question of why do we need to write an sstable.
I guess I'm just not familiar enough with this code to understand what it's doing without further explanations :-(

File: repair/row_level.cc (hunk 2 above)
Comment (mikolajsieluzycki) [reply repair/row_level.cc]: No strong opinion here. I didn't understand the reasoning of calling create_writer multiple times in the loop, while all, apart from the first one, do early exit because the writer was already created. The unit/dtests didn't show any difference. It wasn't intended as performance optimisation. I can definitely remove it.

File: repair/repair_writer.hh (hunk 3 above)
Comment (mikolajsieluzycki) [reply repair/repair_writer.hh:97]: Good point.

File: repair/repair_hash.hh [hunk 12]
```diff
@@ -0,0 +1,50 @@
+#pragma once
+#include <absl/container/btree_set.h>
+#include <cstdint>
+#include <ostream>
+#include "schema.hh"
```
[the remark this answers is not in the document]
Comment (mikolajsieluzycki) [reply repair/repair_hash.hh:5]: It would make the header uncompilable on it's own with error "incomplete type schema used in type trait expression". The hasher has an instance of schema_ptr, which causes the problem. If it was just used in function params, fwd header would be enough.

File: repair/repair_sync_boundary.hh (hunk 4 above)
Comment (mikolajsieluzycki) [reply repair/repair_sync_boundary.hh:16]: Bundling unrelated classes in a cc file can be helpful for compilation times (though it's sometimes complicated to evaluate the gains). Bundling classes in a hh file makes sense only if the classes have to live together (e.g. a container + its iterator) or they're always used together. The latter was the case before my patch, but after splitting the repair_writer to a separate header (for testing), the sync boundary can be either separate or should live with the repair_writer that's actually using it. Should we want to have sync boundary unit tests in the future, it would be better to have a separate header for it instead of having to include repair_writer. In the case of this specific class/file split, it doesn't matter much either way. This becomes problematic when headers organically grow like that and you end up with a header that has 3k lines of code, 50 classes inside and is included by half the codebase (see my readers directory change from a couple weeks back - modifying a super simple reader used in one cc file recompiled a third of scylla).

File: test/boost/repair.cc (hunk 6 above)
Comment (mikolajsieluzycki) [reply test/boost/repair.cc:114]: We're a bit forced here. The goal is to get rid of queue_reader and queue_reader_handle (both are v1 and we already have a v2 version, but can't use it here yet). Repair pipeline uses them internally on the part that the unit tests cover (repair_rows_on_wire -> repair_row -> flush to disk with some extra steps). The goal of this patchset is to write a unit test for that specific part, so that when I get rid of the v1 version of queue reader (in a subsequent patch) I can replace repair_row with a newly written repair_row_v2 and can skip the downgrades.

File: repair/repair_writer.hh (hunk 5 above)
Comment (mikolajsieluzycki) [reply repair/repair_writer.hh]: That's a remnant of my thought process, shouldn't have made it to the patch. The only relevant bit is done a bit later when I remove the parameters from this function and move them to the constructor. Will clean up.

File: test/boost/repair.cc (hunk 7 above)
Comment (mikolajsieluzycki) [reply test/boost/repair.cc:96]: I'll add the explanatory comment to the code, but in the interest of time and clarity let me write it here to get the feedback before I make another push of the patch series.

The test exercises a simplified path of what's happening during regular repair. In normal repair we receive repair_rows_on_wire, they go through various buffers and transformations and at some point are flushed to the disk. There is a lot of work related to figure out which rows to stream, where to pull them from etc. and that code is NOT part of this test.

The preparatory commits extract functions from repair meta that:
- take repair_rows_on_wire and convert them to std::list<repair_row> (repair_meta::to_repair_rows_lsit)
- flush std::list<repair_row> to disk (part of repair_meta::flush_rows_in_working_row_buf)
The test itself effectively just calls those two functions. The fluff around is only to randomly generate inputs and verify that whatever was written to disk is the same as the inputs.

The reason for preparatory work and a somewhat custom writer in the test is that the regular repair uses streaming::make_streaming_consumer, which requires (among other things) sharded<db::system_distributed_keyspace>. That doesn't make much sense in a single node unit test. That "somewhat custom writer" combines 3 functions used widely all over scylla so it's nothing fancy.

I'll reorganise the test code and write comments to make it clearer what the intention is.

With that preamble - I don't think we have to write to disk, though that's what the regular code does. I can rearrange the code to return the flat_mutation_reader_v2 from test_mutation_fragment_queue_impl::push_end_of_stream so that the test compares that reader with the memtable one (which is the source of mutations).

@avikivity As for speed of the test, I think it's acceptable in dev, I haven't checked debug, but will run it shortly.
➜  scylla git:(add-repair-unit-tests/v1) ✗ time ./test.py --mode=dev boost/repair_test
Found 1 tests.
================================================================================
[N/TOTAL]   SUITE    MODE   RESULT   TEST
------------------------------------------------------------------------------
[1/1]       boost     dev   [ PASS ] repair_test.1
------------------------------------------------------------------------------
./test.py --mode=dev boost/repair_test  0.36s user 0.06s system 70% cpu 0.590 total

File: repair/row_level.cc (hunk 8 above)
Comment (mikolajsieluzycki) [reply repair/row_level.cc]: The diff is surprisingly dense, I can update the commit comment to make it more easily digestable.

File: test/boost/repair.cc (hunk 9 above)
Comment (mikolajsieluzycki) [reply test/boost/repair.cc:52]: Will clean up.

File: repair/repair_writer.hh (hunk 3 above)
Comment (xemul) [reply repair/repair_writer.hh:97]: const noexcept qualified

**Comment by xemul** [comment]
LGTM, but I've a generic comment to first five patches. They all introduce headers named `repair/repair_foo.hh`, but what's the point in additional `repair_` prefix if they already sit in the `repair/` directory? How about naming them as `repair/foo.hh`?

**Comment by mikolajsieluzycki** [comment]
@xemul Fair point, will update, though how would you rename repair/repair.hh in this case :grinning: ?

File: test/boost/repair.cc (hunk 7 above)
Comment (denesb) [reply test/boost/repair.cc:96]: Note that we have `cql_env`, which sets up a complete single node environment, including all the system tables. If you need it, feel free to use it. No need to make things more contrived, just to make the unit test a little simpler. We use it in a lot of unit test, when testing code which requires a working database, storage proxy, system tables, etc.

File: repair/row_level.cc [hunk 13]
```diff
@@ -396,45 +396,135 @@ using is_local_reader = bool_class<class is_local_reader_tag>;
...
+        };
+        return sstables::offstrategy(operations_supported.contains(reason));
+    }
+};
+
 future<> repair_writer::write_start_and_mf(lw_shared_ptr<const decorated_key_with_hash> dk, mutation_fragment mf) {
     _current_dk_written_to_sstable = dk;
     if (mf.is_partition_start()) {
-        return _mq->push(std::move(mf)).then([this] {
+        return _impl->queue().push(std::move(mf)).then([this] {
```
[the remark this answers is not in the document]
Comment (mikolajsieluzycki) [reply repair/row_level.cc:457]: Fair point, will store that. The lifetimes of writer and impl are bound anyway, so there is no risk of dangling ref.

**Comment by xemul** [comment]
> @xemul Fair point, will update, though how would you rename repair/repair.hh in this case grinning ?

@mikolajsieluzycki , into `repair/.hh` apparently :trollface:

File: repair/row_level.cc [hunk 14]
```diff
@@ -396,45 +396,135 @@ using is_local_reader = bool_class<class is_local_reader_tag>;
     }
 };
 
+class repair_writer_impl : public repair_writer::impl {
```
[the remark this answers is not in the document]
Comment (mikolajsieluzycki) [reply repair/row_level.cc:399]: In the ideal world (outside of the scope of this series), repair_meta would receive a repair_writer as injected dependency instead of creating it internally. The patch series does NOT instantiate/test repair_meta, so there was no need for that change. That being said, the repair_writer used in test does use test_repair_writer_impl (which inherits from repair_writer::impland test_mutation_fragment_queue_impl (which inherits from mutation_fragment_queue::impl). 

repair_writer::impl was abstracted to hide the process of creating the underlying writer (multishard vs the test one) and waiting until it's finished writing. The queue itself is a separate concept as it exposes an api to write to it and hides the details of the consumer of that queue (writing to sstables in production code vs the test one).

In the pre-patch case you could test what the sstable has written to the disk, but you still need a channel to provide the relevant mutations. This patch series is written with the subsequent change in mind, which changes the internal representation of repair_row (which contains mutation_fragment) to repair_row_v2 (which contains mutation_fragment_v2). The whole path then takes repair_rows_on_wire (contains mutation_fragment v1) converts to internal representation and then flushes to disk. There is no point in testing the repair_row (or repair_row_v2) itself, as this is an implementation detail. We need to ensure the path of repair_rows_on_wire to disk is executed properly, regardless of the implementation details.

File: test/boost/repair.cc (hunk 7 above)
Comment (mikolajsieluzycki) [reply test/boost/repair.cc:96]: I actually started with that, but it doesn't provide _system_distributed_keyspace, needed to create the streaming consumer that writes the mutations to sstables. After talking with @bhalevy, he mentioned that until we have multi-node test environments it doesn't make sense to have that distributed keyspace in cql_test_env (unless I'm misquoting). Maybe I could refactor and abstract away the usage of the keyspace in the consumer, but that sounded like avoiding one rabbit hole by going head first into another one.

File: repair/repair_hash.hh (hunk 1 above)
Comment (mikolajsieluzycki) [reply repair/repair_hash.hh:45]: Will change, thanks.

File: test/boost/repair.cc (hunk 7 above)
Comment (denesb) [reply test/boost/repair.cc:96]: I think there is nothing preventing us from adding distributed system keyspace to `cql_test_env`, just like we have it set-up on single-node clusters. Don't know how trivial that is though.

Comment (denesb) [reply test/boost/repair.cc:96]: We actually already have it: https://github.com/scylladb/scylla/blob/59d56a3fd7ab22a3184de2ebfb3df67580316ff9/test/lib/cql_test_env.cc#L716
It is just not exposed directly.

File: repair/repair_writer.hh (hunk 3 above)
Comment (mikolajsieluzycki) [reply repair/repair_writer.hh]: Fixed.

File: test/boost/repair.cc (hunk 7 above)
Comment (mikolajsieluzycki) [reply test/boost/repair.cc]: Reorganised the code and added comments

Comment (mikolajsieluzycki) [reply test/boost/repair.cc]: Reorganised the test code, added comments and removed flushing in favour of comparing memtable reader + reader from fragments that would have been passed to sstable::write_components.

File: test/boost/repair.cc (hunk 9 above)
Comment (mikolajsieluzycki) [reply test/boost/repair.cc]: Fixed

File: test/boost/repair.cc (hunk 11 above)
Comment (mikolajsieluzycki) [reply test/boost/repair.cc]: Added

**Comment by avikivity** [comment]
What kind of bugs will this test find? e.g. show some bug that could have happened in the repair code and failed this test.

**Comment by mikolajsieluzycki** [comment]
The main point of this test is having a baseline unit test for current behaviour. A subsequent patch series (which I'm currently working on) will switch parts of repair code from using repair_row to repair_row_v2 (to get rid of queue_reader and have only queue_reader_v2). When I have a baseline test like this one, I can just introduce one conversion function, sprinkle the code with v2 suffixes and check if everything works in a couple seconds. It already showed me bugs in my repair_row to repair_row_v2 conversion without having to run dtest (simplest repair-related test takes 6minutes there), which was the main point of this change.

**Comment by denesb** [comment]
@mikolajsieluzycki for future reference, when force-pushing a new version of a PR, please add a comment with the changelog, so reviewers can focus their attention only on the changed areas, instead of having to re-review everything.

**Comment by mikolajsieluzycki** [comment]
@denesb The changed areas are only the ones the reviewers asked to change, but I'll make a note to do this explictly in the PR description in the future.

**Review by denesb — APPROVED** [review]

**Comment by denesb** [comment]
> @denesb The changed areas are only the ones the reviewers asked to change, but I'll make a note to do this explictly in the PR description in the future.

Yes, I did go though the comments to collect what changed, but that is harder to do than reading a nicely prepared changelog. Also, comments sometimes disappear as the code is updated.
